python-spec/src/somacore/coordinates.py

```diff
@@ -0,0 +1,453 @@
+        """The names of the axes of the input coordinate space.
+
+        Lifecycle: experimental
+        """
+        return self._input_axes
+
+    @property
+    def output_axes(self) -> Tuple[str, ...]:
+        """The names of the axes of the output coordinate space.
+
+        Lifecycle: experimental
+        """
+        return self._output_axes
+
+
+class AffineTransform(CoordinateTransform):
+    """An affine coordinate trasformation from one coordinate space to another.
+
+    An affine transform is a combination of a linear transformation and a translation.
+
+    Args:
+        input_axes: The names of the axes for the input coordinate space.
+        output_axes: The names of the axes for the output coordinate space.
+        matrix: Matrix (perhaps augmented) that represents the affine transformation.
+            Can be provided as just the linear transform (if no translation), the
+            full augmented matrix, or the augmented matrix without the final row.
+
+    Lifecycle: experimental
+    """
+
+    def __init__(
+        self,
+        input_axes: Union[str, Sequence[str]],
+        output_axes: Union[str, Sequence[str]],
+        matrix: npt.ArrayLike,
+    ):
+        super().__init__(input_axes, output_axes)
+
+        # Check the rank of the input/output axes match.
+        if len(self.input_axes) != len(self.output_axes):
+            raise ValueError(
+                "The input axes and output axes must be the same length for an "
+                "affine transform."
+            )
+        rank = len(self.input_axes)
+
+        # Create and validate the augmented matrix.
+        self._matrix: npt.NDArray[np.float64] = np.array(matrix, dtype=np.float64)
+        if self._matrix.shape == (rank + 1, rank + 1):
+            if not (
+                self._matrix[-1, -1] == 1.0
+                and np.array_equal(self._matrix[-1, :-1], np.zeros((rank,)))
+            ):
+                raise ValueError(
+                    f"Input matrix {self._matrix} has augmented matrix shape, but is not a valid "
+                    f"augmented matrix."
+                )
+        elif self._matrix.shape == (rank, rank + 1):
+            self._matrix = np.vstack(
+                (
+                    self._matrix,
+                    np.hstack((np.zeros((rank,)), np.array([1]))),
+                )
+            )
+        elif self._matrix.shape == (rank, rank):
+            self._matrix = np.vstack(
+                (
+                    np.hstack((self._matrix, np.zeros((rank, 1)))),
+                    np.hstack((np.zeros((rank,)), np.array([1]))),
+                )
+            )
+        else:
+            raise ValueError(
+                f"Unexpected shape {self._matrix.shape} for the input affine matrix."
+            )
+
+    def __matmul__(self, other: Any) -> CoordinateTransform:
+        if self.input_axes != other.output_axes:
+            raise ValueError("Axis mismatch between transformations.")
```

Comment:
Probably also worth adding this as something similar to `_check_same_length` above, so that you don’t need to repeat these lines (and you can maybe give the user a slightly better error message, like "output axes of {self_type} must match input axes of {other_type}"

Reply:
you missed removing this if-statement!